src/composer.rs

```diff
@@ -129,7 +129,7 @@ pub trait Composer: Sized + Index<Witness, Output = BlsScalar> {
     ) -> Witness {
         // the bits are iterated as chunks of two; hence, we require an even
         // number
-        debug_assert_eq!(num_bits & 1, 0);
+        assert_eq!(num_bits & 1, 0, "number of bits must be even");
```

Comment:
Not sure if it makes sense to add production panic to check circuit development consistency. I'd rather keep it as `debug_assert_eq` and have tests checking for such case

Reply:
I have the same feeling on panicking like this, but more because it signals that we should be erroring rather than panicking. The library should validate user input against what it expects.

Reply:
There is an important distinction tho: the validation is static and should be performed at compilation time.

Meaning: when we create a circuit implementation, we create it for a static type (i.e. `impl Circuit for Foo`). It means it could be a waste to move such check to runtime (i.e. make the function fallible & `Err`).

If we move everything to static time, then it probably makes more sense. In short: if the function receives a constant, we can just `assert` it and it would fail at compile time if the user attempts to create a circuit with wrong bits boundaries:

```rust
fn append_logic_component<BITS: usize>(
        &mut self,
        a: Witness,
        b: Witness,
        is_component_xor: bool,
    ) -> Witness {
    assert_eq!(BITS & 1, 0, "some msg");
```

The code above will not compile for invalid bits.

Reply:
In fact, the problem above probably applies elsewhere. We should probably use more static stuff for other components instead of running test checks. Probably `is_component_xor` should also be a constant.

TLDR; components should receive only witnesses as arguments. Everything else is a constant

Reply:
We can also circumvent any misuse if we let the user pass the amount of bit-pairs instead of bits. In the components themselves we multiply the bit-pairs by 2 and can therefore always be sure the total amount of bits is dividable by two. 

It would be more cognitive overhead on the circuit creator but would make sure the components can not be misused without the need to panic or error

Reply:
I did an attempt on my above proposal combined with @vlopes11 suggestion of making the logic and range components generic over the bit (bit-halves in this case).

@vlopes11 and @ureeves please check it out under the branch: [mocello/763_fix_debug_panic_bit_pairs](https://github.com/dusk-network/plonk/tree/mocello/763_fix_debug_panic_bit_pairs) and let me know whether you like that one or the latest changes on this branch more.

Reply:
I very much like the direction of "making illegal state unrepresentable". Comes with tons of benefits. I vote for the [`bit_pairs`](https://github.com/dusk-network/plonk/tree/mocello/763_fix_debug_panic_bit_pairs) branch that you suggested.

Reply:
I also like the solution with the `BIT_PAIRS` most.

Pushed to the branch, please have a final look and approve (*if* you approve that is ;) )